File: examples/agent-framework-webhooks/python/agent_webhooks/handler.py
"""Framework-neutral handling for verified e2a webhook deliveries."""

from __future__ import annotations

from e2a import construct_event

from .contracts import InboundResource, ReplyAgent
from .delivery_state import EventDeduper


class DeliveryInProgress(Exception):
    """Raised when another request is already processing the event."""

    def __init__(self, event_id: str) -> None:
        self.event_id = event_id
        super().__init__(f"delivery {event_id} is already in progress")


async def handle_delivery(
    body: bytes,
    signature: str,
    secret: str,
    inbound: InboundResource,
    agent: ReplyAgent,
    deduper: EventDeduper,
) -> dict[str, str]:
    """Verify, claim, and process one webhook delivery."""

    event = construct_event(body, signature, secret)
    if event.type != "email.received":
        return {"status": "ignored"}

    claim = await deduper.claim(event.id)
    if claim == "processed":
        return {"status": "duplicate"}
    if claim == "processing":
        raise DeliveryInProgress(event.id)

    try:
        email = await inbound.from_event(event)
        reply_text = (await agent.reply(email)).strip()
        if not reply_text:
            await deduper.complete(event.id)
            return {"status": "no_reply", "conversation_id": email.conversation_id}

        result = await email.reply(
            {"text": reply_text, "conversation_id": email.conversation_id},
            idempotency_key=event.id,
        )
        await deduper.complete(event.id)
        status = "replied" if result.status in {"accepted", "sent"} else result.status
        return {"status": status, "conversation_id": email.conversation_id}
    except BaseException:
        await deduper.release(event.id)
        raise

File: examples/agent-framework-webhooks/typescript/src/handler.ts
import { constructEvent, type SendResultView } from "@e2a/sdk/v1";

import type { InboundResource, ReplyAgent } from "./contracts.js";
import { EventDeduper, conversationIdFor } from "./delivery-state.js";

export class DeliveryInProgress extends Error {
  readonly eventId: string;

  constructor(eventId: string) {
    super(`delivery ${eventId} is already in progress`);
    this.name = "DeliveryInProgress";
    this.eventId = eventId;
  }
}

export interface HandleDeliveryOptions {
  body: string | Uint8Array;
  signature: string;
  secret: string;
  inbound: InboundResource;
  agent: ReplyAgent;
  deduper: EventDeduper;
}

export type HandleDeliveryResult =
  | { kind: "ignored"; status: "ignored" }
  | { kind: "duplicate"; status: "duplicate" }
  | { kind: "no_reply"; status: "no_reply"; conversationId: string }
  | { kind: "replied"; status: "replied"; conversationId: string }
  | { kind: "send_result"; status: string; conversationId: string };

/** Verify, claim, and process one webhook delivery. */
export async function handleDelivery({
  body,
  signature,
  secret,
  inbound,
  agent,
  deduper,
}: HandleDeliveryOptions): Promise<HandleDeliveryResult> {
  const rawBody = typeof body === "string" ? body : Buffer.from(body.buffer, body.byteOffset, body.byteLength);
  const event = constructEvent(rawBody, signature, secret);
  if (event.type !== "email.received") return { kind: "ignored", status: "ignored" };

  const claim = await deduper.claim(event.id);
  if (claim === "processed") return { kind: "duplicate", status: "duplicate" };
  if (claim === "processing") throw new DeliveryInProgress(event.id);

  try {
    const email = await inbound.fromEvent(event);
    const conversationId = conversationIdFor(event.id, email.conversationId);
    const replyText = (await agent.reply(email, conversationId)).trim();
    if (replyText.length === 0) {
      await deduper.complete(event.id);
      return { kind: "no_reply", status: "no_reply", conversationId };
    }

    const result: SendResultView = await email.reply(
      { text: replyText, conversationId },
      { idempotencyKey: event.id },
    );
    await deduper.complete(event.id);
    const status = result.status === "accepted" || result.status === "sent" ? "replied" : result.status;
    return status === "replied"
      ? { kind: "replied", status, conversationId }
      : { kind: "send_result", status, conversationId };
  } catch (error: unknown) {
    await deduper.release(event.id);
    throw error;
  }
}
